### PR TITLE
convert JSDoc typedef to type, issue 50644

### PR DESCRIPTION
**Pull request by brendaHuang**
Fixes https://github.com/microsoft/TypeScript/issues/50644

This PR converts JSDoc typedef to type. It converts two common use cases of @typedef as listed in [the JSDoc documentation](https://jsdoc.app/tags-typedef.html) and other variations, such as nested object types and using @typedef to define primitive types. 

Of the five test files, one one addresses the union type, and one primitive type.
The other three addresses object types (index signatures), including nested object types and object types that have comments attached to the properties.

### CODE REVIEW

**Comment by brendaHuang** [comment]
@microsoft-github-policy-service agree [company="{LinkedIn}"]

**Comment by brendaHuang** [comment]
@microsoft-github-policy-service agree company="LinkedIn"

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 1]
```diff
@@ -9834,6 +9834,11 @@ declare namespace ts {
         emit(writeFile?: WriteFileCallback, customTransformers?: CustomTransformers): EmitResult | BuildInvalidedProject<T> | undefined;
     }
     type InvalidatedProject<T extends BuilderProgram> = UpdateOutputFileStampsProject | BuildInvalidedProject<T> | UpdateBundleProject<T>;
+    namespace codefix {
+        function _containsJSDocTypedef(node: Node): node is HasJSDoc;
+        function getJSDocTypedefNode(node: HasJSDoc): JSDocTypedefTag;
+        function containsTypeDefTag(node: Node): node is JSDocTypedefTag;
+    }
```
Comment (jakebailey) [inline tests/baselines/reference/api/tsserverlibrary.d.ts:9841]: This indicates that you are exporting things but not marking them as internal. I would look at other code fixes for examples.

Comment (brendaHuang) [reply tests/baselines/reference/api/tsserverlibrary.d.ts]: @jakebailey Fixed.

**Comment by brendaHuang** [comment]
@andrewbranch Would you take a look at this PR? Thanks.

File: src/services/codefixes/convertTypedefToType.ts [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+            return factory.createTypeReferenceNode("object");
+        case SyntaxKind.VoidKeyword:
+            return factory.createTypeReferenceNode("void");
+        default:
+            return;
+    }
+}
+
+/** @internal */
+export function _containsJSDocTypedef(node: Node): node is HasJSDoc {
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:208]: Style nit, but you don't need the underscore prefix here (as far as I'm concerned).

File: src/services/codefixes/convertTypedefToType.ts [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    if (hasJSDocNodes(node)) {
+        const jsDocNodes = node.jsDoc || [];
+        return jsDocNodes.some((node) => {
+            const tags = node.tags || [];
+            return tags.some((tag) => isJSDocTypedefTag(tag));
+        });
+    }
+    return false;
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:216]: We have helpers for this sort of pattern; you probably could just write something like:

```ts
return hasJSDocNode(node) && some(node.jsDoc, node => some(node.tags, tag => isJSDocTypedefTag(tag)))
```

File: src/services/codefixes/convertTypedefToType.ts [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    );
+    return declaration;
+}
+
+function createSignatureFromTypeLiteral(typeLiteral: JSDocTypeLiteral): PropertySignature[] | undefined {
+    const propertyTags = typeLiteral.jsDocPropertyTags;
+    if(!propertyTags || propertyTags.length === 0) return;
+    const props = (propertyTags as JSDocPropertyTag[]).reduce((
+        signatures: PropertySignature[], tag: JSDocPropertyTag
+        ) => {
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:121]: Generally, you want to use the `reduceLeft` helper instead, which is a lot faster.

File: src/services/codefixes/convertTypedefToType.ts [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        factory.createIdentifier(typeName),
+        [],
+        typeReference
+    );
+    return declaration;
+}
+
+function createSignatureFromTypeLiteral(typeLiteral: JSDocTypeLiteral): PropertySignature[] | undefined {
+    const propertyTags = typeLiteral.jsDocPropertyTags;
+    if(!propertyTags || propertyTags.length === 0) return;
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:118]: nit: spacing after if here.

File: src/services/codefixes/convertTypedefToType.ts [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+            [],
+            propertySignatures,
+        );
+        return interfaceDeclaration;
+}
+
+function createTypeAliasForTypeExpression(
+    typeName: string,
+    typeExpression: JSDocTypeExpression
+    ): TypeAliasDeclaration | undefined {
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:104]: nit: stray indent

File: src/services/codefixes/convertTypedefToType.ts [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    // For use case @typedef {object}Foo @property{bar}number
+    // But object type can be nested, meaning the value in the k/v pair can be object itself
+    if (typeExpression.kind === SyntaxKind.JSDocTypeLiteral) {
+        return createInterfaceForTypeLiteral(typeName, typeExpression);
+    }
+    // for use case @typedef {(number|string|undefined)} Foo or @typedef {number|string|undefined} Foo
+    // In this case, we reach the leaf node of AST.
+    // Here typeExpression.type is a TypeNode, e.g. a UnionType or a primitive type, such as "number".
+    if (typeExpression.kind === SyntaxKind.JSDocTypeExpression) {
+     return createTypeAliasForTypeExpression(typeName, typeExpression);
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:81]: nit: bad indent

There are actually a lot of these spacing problems in the file; I'll stop commenting on them as I wish we had a formatter to prevent this, but I would take a look at the diff and just see what's missing.

File: src/services/codefixes/convertTypedefToType.ts [hunk 8]
```diff
@@ -0,0 +1,232 @@
+import {
+    Diagnostics, factory, SourceFile, SyntaxKind, Node, JSDocTypedefTag,
+    JSDocTypeLiteral, JSDocPropertyTag, JSDocTypeExpression, InterfaceDeclaration,
+    TypeAliasDeclaration, ParenthesizedTypeNode, UnionTypeNode, PropertySignature,
+    TypeNodeSyntaxKind, TypeNode, HasJSDoc, isJSDocTypedefTag, hasJSDocNodes,
+    isInJSDoc, getTokenAtPosition, textChanges, flatMap
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:6]: Please make these a big vertical list and sorted (may need to merge main to get linting for this); otherwise we'll end up with merge conflicts and lint errors.

File: src/compiler/diagnosticMessages.json [hunk 9]
```diff
@@ -7548,5 +7548,13 @@
...
         "category": "Error",
         "code": 18050
+    },
+    "Convert typedef to type": {
+        "category": "Message",
+        "code": 18051
+    },
+    "Convert all typedefs to types": {
+        "category": "Message",
+        "code": 18052
```
Comment (jakebailey) [inline src/compiler/diagnosticMessages.json:7558]: I think these should be in the 8000 block (see the codes used for the other diagnostics added in the part of the suggestionDiagnostics file you're modifying.

**Review by jakebailey — CHANGES_REQUESTED** [review]
This is a first past skim review; I think I'd need to think of some edge cases here. Probably @andrewbranch and @sandersn know more.

File: src/services/codefixes/convertTypedefToType.ts (hunk 5 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts:118]: Fixed

File: src/services/codefixes/convertTypedefToType.ts (hunk 6 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts:104]: fixed

File: src/services/codefixes/convertTypedefToType.ts (hunk 2 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts:208]: fixed

Comment (jakebailey) [reply src/services/codefixes/convertTypedefToType.ts:208]: Are you marking these fixed as you fix them locally to push later, or did you intend to push? (I'm not seeing pushes, hence me asking.)

**Comment by jakebailey** [comment]
It's too late now, but if you can avoid it, I'd appreciate merges from main and additional commits (instead of force pushing) as it's hard to follow what changed in the PR and what comments have been addressed. We squash on PR merge so no need to keep the PR clean. (See our contributing guide: https://github.com/microsoft/TypeScript/blob/main/CONTRIBUTING.md#force-pushing)

File: src/services/codefixes/convertTypedefToType.ts (hunk 4 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed

File: src/services/codefixes/convertTypedefToType.ts (hunk 3 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed

File: src/services/codefixes/convertTypedefToType.ts (hunk 7 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed

File: src/services/codefixes/convertTypedefToType.ts (hunk 8 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed

File: src/compiler/diagnosticMessages.json (hunk 9 above)
Comment (brendaHuang) [reply src/compiler/diagnosticMessages.json]: fixed

File: src/services/codefixes/convertTypedefToType.ts [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+                name,
+                // eslint-disable-next-line local/boolean-trivia
+                undefined,
+                typeReference
+            );
+            return [...signatures, prop];
+        }
+    };
+
+    const props = reduceLeft(propertyTags, getSignatures, []);
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:166]: Could this just be map or mapDefined?

File: tests/cases/fourslash/codeFixConvertTypedefToType4.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+////
+//// /**
+////  * @typedef {object} Person
+////  * @property {object} data
+////  * @property {string} data.name
+////  * @property {number} data.age
+////  * @property {object} data.contact
+////  * @property {string} data.contact.address
+////  * @property {string} data.contact.phone
```
Comment (jakebailey) [inline tests/cases/fourslash/codeFixConvertTypedefToType4.ts:11]: Can you test something like `[data.contact.address]`, which should produce an optional property?

**Comment by brendaHuang** [comment]
> It's too late now, but if you can avoid it, I'd appreciate merges from main and additional commits (instead of force pushing) as it's hard to follow what changed in the PR and what comments have been addressed. We squash on PR merge so no need to keep the PR clean. (See our contributing guide: https://github.com/microsoft/TypeScript/blob/main/CONTRIBUTING.md#force-pushing)

Okay.

File: src/services/codefixes/convertTypedefToType.ts [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+        }
+        // Leaf node, where type.kind === SyntaxKind.JSDocTypeExpression
+        else if (type) {
+            typeReference = createTypeReference(type);
+        }
+        if (typeReference && name) {
+            const prop = factory.createPropertySignature(
+                [],
+                name,
+                // eslint-disable-next-line local/boolean-trivia
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:158]: You should fix this lint rather than hiding it; e.g. put `/*nameOfParameter*/` before `undefined` to indicate what parameter it's intending to be.

Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed

File: tests/cases/fourslash/codeFixConvertTypedefToType4.ts (hunk 11 above)
Comment (brendaHuang) [reply tests/cases/fourslash/codeFixConvertTypedefToType4.ts]: done

File: src/services/codefixes/convertTypedefToType.ts (hunk 10 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: changed to mapDefined

File: src/services/codefixes/convertTypedefToType.ts [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+        return createTypeAliasForTypeExpression(typeName, typeExpression);
+    }
+}
+
+function createInterfaceForTypeLiteral(
+    typeName: string,
+    typeLiteral: JSDocTypeLiteral
+): InterfaceDeclaration | undefined {
+    const propertySignatures = createSignatureFromTypeLiteral(typeLiteral);
+    if (!propertySignatures || propertySignatures.length === 0) return;
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:110]: This can just be `if (!some(propertySignatures)) return`.

File: src/services/codefixes/convertTypedefToType.ts [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+        factory.createIdentifier(typeName),
+        [],
+        typeReference
+    );
+    return declaration;
+}
+
+function createSignatureFromTypeLiteral(typeLiteral: JSDocTypeLiteral): PropertySignature[] | undefined {
+    const propertyTags = typeLiteral.jsDocPropertyTags;
+    if (!propertyTags || propertyTags.length === 0) return;
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:138]: Same here, could just be `if (!some(propertyTags)) return`.

File: src/services/codefixes/convertTypedefToType.ts [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+    const getSignature = (tag: JSDocPropertyTag) => {
+        const name = getPropertyName(tag);
+        const type = tag.typeExpression?.type;
+        const isOptional = tag.isBracketed;
+        let typeReference;
+
+         // Recursively handle nested object type
+         if (type && type.kind === SyntaxKind.JSDocTypeLiteral) {
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:147]: Spacing error here.
```suggestion
        if (type && type.kind === SyntaxKind.JSDocTypeLiteral) {
```

**Review by jakebailey — APPROVED** [review]
I don't see anything wrong at this point; the tests appear to show that it works. Someone else may have some comments, though.

File: src/services/codefixes/convertTypedefToType.ts (hunk 13 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed

File: src/services/codefixes/convertTypedefToType.ts (hunk 14 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed

File: src/services/codefixes/convertTypedefToType.ts (hunk 15 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed

File: src/services/codefixes/convertTypedefToType.ts [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+        const changes = textChanges.ChangeTracker.with(context, t => doChange(t, node, context.sourceFile));
+
+        if (changes.length > 0) {
+            return [
+                createCodeFixAction(
+                    fixId,
+                    changes,
+                    Diagnostics.JSDoc_typedef_may_be_converted_to_TypeScript_type,
+                    fixId,
+                    Diagnostics.JSDoc_typedefs_may_be_converted_to_TypeScript_types
```
Comment (andrewbranch) [inline src/services/codefixes/convertTypedefToType.ts:50]: These are the texts that show up in the quick fix menu and should use a different pair of diagnostics than the suggestion, phrased as an imperative (“Convert typedef to TypeScript type”).

See the `require` to `import` fix for example:

image


image

File: src/services/codefixes/convertTypedefToType.ts [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+
+function getPropertyName(tag: JSDocPropertyTag): string | undefined {
+    const { name } = tag;
+    if (!name) return;
+
+    let propertyName;
+    // for "@property {string} parent.child" or "@property {string} parent.child.grandchild" in nested object type
+    // We'll get "child" in the first example or "grandchild" in the second example as the prop name
+    if (name.kind === SyntaxKind.QualifiedName) {
+        propertyName = name.right.getFullText().trim();
```
Comment (andrewbranch) [inline src/services/codefixes/convertTypedefToType.ts:181]: Everywhere you use `getFullText`:

```suggestion
        propertyName = name.right.text;
```

File: src/services/codefixes/convertTypedefToType.ts [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+    // Create TypeReferenceNode for UnionType
+    if (type.kind === SyntaxKind.UnionType) {
+        const elements = (type as UnionTypeNode).types;
+        const nodes = mapDefined(
+            elements,
+            (element) => {
+                const node = transformUnionTypeKeyword(element.kind);
+                if (node) {
+                    return node;
+                }
```
Comment (andrewbranch) [inline src/services/codefixes/convertTypedefToType.ts:204]:
```suggestion
                return transformUnionTypeKeyword(element.kind);
```

File: src/services/codefixes/convertTypedefToType.ts [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+        case SyntaxKind.NumberKeyword:
+            return factory.createTypeReferenceNode("number");
+        case SyntaxKind.StringKeyword:
+            return factory.createTypeReferenceNode("string");
+        case SyntaxKind.UndefinedKeyword:
+            return factory.createTypeReferenceNode("undefined");
+        case SyntaxKind.ObjectKeyword:
+            return factory.createTypeReferenceNode("object");
+        case SyntaxKind.VoidKeyword:
+            return factory.createTypeReferenceNode("void");
```
Comment (andrewbranch) [inline src/services/codefixes/convertTypedefToType.ts:229]: Where did this set come from? It’s not obvious to me why other intrinsic types are excluded, like `boolean`, `bigint`, `null`. Also, what do these have to do with union types?

File: src/services/codefixes/convertTypedefToType.ts [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+        propertyName = name.right.getFullText().trim();
+    }
+    else {
+        propertyName = tag.name.getFullText().trim();
+    }
+    return propertyName;
+}
+
+// Create TypeReferenceNode when we reach the leaf node of AST
+function createTypeReference(type: TypeNode): TypeNode | undefined {
```
Comment (andrewbranch) [inline src/services/codefixes/convertTypedefToType.ts:190]: This function seems leave out more than it covers (intersections, object types, type references, conditional types, etc.). Why can’t we just clone the type node that we got from the JSDoc type expression?

**Review by andrewbranch — CHANGES_REQUESTED** [review]
I haven’t looked at the code that converts `JSDocTypeLiteral`s, but I think that’s the only code that should exist—there’s a lot of deep walking and transforming existing TypeScript syntax that I think can just be replaced with a `getSynthesizedDeepClone(typeExpression)`.

File: src/services/codefixes/convertTypedefToType.ts (hunk 16 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: changed

File: src/services/codefixes/convertTypedefToType.ts (hunk 17 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: changed

File: src/services/codefixes/convertTypedefToType.ts (hunk 19 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: removed function

File: src/services/codefixes/convertTypedefToType.ts (hunk 20 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: done

File: src/services/codefixes/convertTypedefToType.ts [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+
+    const getSignature = (tag: JSDocPropertyTag) => {
+        const name = getPropertyName(tag);
+        const type = tag.typeExpression?.type;
+        const isOptional = tag.isBracketed;
+        let typeReference;
+
+        // Recursively handle nested object type
+        if (type && type.kind === SyntaxKind.JSDocTypeLiteral) {
+            const signatures = createSignatureFromTypeLiteral(type as JSDocTypeLiteral);
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:144]: I don't quite understand these "as" uses; it seems like the functions should be changing rather than throwing away safety.

Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts:144]: Please let me know how to change the function. As Line 143 shows, the argument is JSDocTypeLiteral and also is a type node. Removing the 'as' gives an error.

**Comment by brendaHuang** [comment]
@jakebailey @andrewbranch The tests passed locally. I don't know why they failed in CI. Please let me know how to fix them. Thank you for your help.

File: src/services/codefixes/convertTypedefToType.ts (hunk 21 above)
Comment (jakebailey) [reply src/services/codefixes/convertTypedefToType.ts:144]: Maybe it's the other "as", but I was referring to the compiler errors distinguishing JSDocPropertyTag from a variant with "Like" in the name. This one is fine.

**Comment by jakebailey** [comment]
When's the last time you merged main? It's possible that CI is testing the merge head and so has extra code not on your branch.

File: src/services/codefixes/convertTypedefToType.ts (hunk 21 above)
Comment (jakebailey) [reply src/services/codefixes/convertTypedefToType.ts:144]: I gave the above feedback while not actually at my computer so I didn't go into enough detail (sorry); I think here you could just use `isJSDocTypeLiteral` and avoid the cast, but this one is very low priority as it's very common not to use these guards in the codebase for perf reasons (but I don't think this part is perf critical).

File: src/services/codefixes/convertTypedefToType.ts [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+                name,
+                questionToken,
+                typeReference
+            );
+
+            return prop;
+        }
+    };
+
+    const props = mapDefined(propertyTags as JSDocPropertyTag[], getSignature);
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:165]: This is the cast I didn't like; `getSignature` is defined in this file and should instead take a `JSDocPropertyLikeTag` + fixes to make it work, as there are multiple nodes that implement that interface.

**Review by jakebailey — CHANGES_REQUESTED** [review]

File: src/services/codefixes/convertTypedefToType.ts [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+    return flatMap(jsDocNodes, (node) => {
+        const tags = node.tags || [];
+        return tags.filter((tag) => isJSDocTypedefTag(tag));
+    })[0] as unknown as JSDocTypedefTag;
+}
+
+/** @internal */
+export function containsTypeDefTag(node: Node): node is JSDocTypedefTag {
+    return isInJSDoc(node) && isJSDocTypedefTag(node);
+}
```
Comment (andrewbranch) [inline src/services/codefixes/convertTypedefToType.ts:204]: This function seems to be named incorrectly, and isn’t it redundant with just `isJSDocTypedefTag`? I didn’t think it was possible for something to pass `isJSDocTypedefTag` without implying `isInJSDoc`.

File: src/services/codefixes/convertTypedefToType.ts [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+        propertyName = tag.name.getText();
+    }
+    return propertyName;
+}
+
+
+/** @internal */
+export function containsJSDocTypedef(node: Node): node is HasJSDoc {
+    return hasJSDocNodes(node) && some(node.jsDoc, node => some(node.tags, tag => isJSDocTypedefTag(tag)));
+}
```
Comment (andrewbranch) [inline src/services/codefixes/convertTypedefToType.ts:189]: Suggestion: delete this function and use `getJSDocTypedefNode` instead. They both do the same work, and you’ll know if a node `containsJSDocTypeDef` by whether `getJSDocTypedefNode` returns undefined or not. And then once you know, you’ll already have the node instead of having to fish it out again.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

File: src/services/codefixes/convertTypedefToType.ts [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+    let propertyName;
+    // for "@property {string} parent.child" or "@property {string} parent.child.grandchild" in nested object type
+    // We'll get "child" in the first example or "grandchild" in the second example as the prop name
+    if (name.kind === SyntaxKind.QualifiedName) {
+        propertyName = name.right.text;
+    }
+    else {
+        propertyName = tag.name.getText();
+    }
+    return propertyName;
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/services/codefixes/convertTypedefToType.ts:182]: Alternatively, this probably doesn’t need to be in a function at all. You can just inline this where it’s called.

File: src/services/codefixes/convertTypedefToType.ts [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+
+/** @internal */
+export function getJSDocTypedefNode(node: HasJSDoc): JSDocTypedefTag {
+    const jsDocNodes = node.jsDoc || [];
+
+    return flatMap(jsDocNodes, (node) => {
+        const tags = node.tags || [];
+        return tags.filter((tag) => isJSDocTypedefTag(tag));
+    })[0] as unknown as JSDocTypedefTag;
+}
```
[the remark this answers is not in the document]
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: changed

File: src/services/codefixes/convertTypedefToType.ts (hunk 24 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: deleted

File: src/services/codefixes/convertTypedefToType.ts (hunk 23 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: changed

File: src/services/codefixes/convertTypedefToType.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: changed

File: src/services/codefixes/convertTypedefToType.ts (hunk 22 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed as suggested

File: src/services/codefixes/convertTypedefToType.ts (hunk 21 above)
Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: fixed as suggested

**Comment by brendaHuang** [comment]
@jakebailey @andrewbranch After I rebased yesterday, the newly added tests failed. As I ran the debugger, the type alias declaration was successfully created. Then node.modifiers is passed into [Line 1280 of src/compiler/visitorPublic.ts](https://github.com/microsoft/TypeScript/blob/main/src/compiler/visitorPublic.ts#L1280), then [assignPositionsToNodeArray() of src/services/textChanges.ts](https://github.com/microsoft/TypeScript/blob/main/src/services/textChanges.ts#L1345), then eventually failed at [Line 178 in the getPos() function of src/services/textChanges.ts](https://github.com/microsoft/TypeScript/blob/main/src/services/textChanges.ts#L178). The value of node.modifiers is [pos: -1, end: -1, hasTrailingComma: false, transformFlags: 0]. Please see screenshots attached. Thank you very much for your help!

Screenshot 2023-02-26 at 7 25 33 AM
Screenshot 2023-02-26 at 7 29 00 AM
Screenshot 2023-02-26 at 7 34 14 AM
Screenshot 2023-02-26 at 7 58 02 AM

**Comment by jakebailey** [comment]
I checked, and it's because you're passing in an empty array for modifiers to `createTypeAliasDeclaration`. I would recheck all of your `create` functions to see if you could actually just pass in `undefined` instead.

**Comment by brendaHuang** [comment]
@jakebailey Thank you very much! The tests now passed. Please review again.

File: src/services/codefixes/convertTypedefToType.ts [hunk 27]
```diff
@@ -104,10 +104,10 @@ function createInterfaceForTypeLiteral(
     const propertySignatures = createSignatureFromTypeLiteral(typeLiteral);
     if (!some(propertySignatures)) return;
     const interfaceDeclaration = factory.createInterfaceDeclaration(
-        undefined,
+        /*nameOfParameter*/ undefined,
```
Comment (jakebailey) [inline src/services/codefixes/convertTypedefToType.ts:107]: All of these comments should actually name the parameter they correspond to.

Comment (brendaHuang) [reply src/services/codefixes/convertTypedefToType.ts]: Haha. That was a silly mistake I made. Thanks. Changed.

**Comment by brendaHuang** [comment]
@jakebailey @andrewbranch Does this PR look OK? Thanks.

**Review by jakebailey — APPROVED** [review]
Seems reasonable, I think? Still not an expert, but it seems about as concise as I can imagine.

**Comment by brendaHuang** [comment]
@andrewbranch Would you take a look? Thank you very much.

**Review by andrewbranch — APPROVED** [review]
Nice work, thanks @brendaHuang!
